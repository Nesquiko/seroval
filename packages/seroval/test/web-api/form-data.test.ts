import { describe, it, expect } from 'vitest';
import {
  crossSerialize,
  crossSerializeAsync,
  crossSerializeStream,
  deserialize,
  fromJSON,
  serialize,
  serializeAsync,
  toJSON,
  toJSONAsync,
} from '../../src';

describe('FormData', () => {
  describe('serialize', () => {
    it('supports FormData', () => {
      const example = new FormData();
      example.set('hello', 'world');
      example.set('foo', 'bar');
      const result = serialize(example);
      expect(result).toMatchSnapshot();
      const back = deserialize<FormData>(result);
      expect(back).toBeInstanceOf(FormData);
    });
  });
  describe('serializeAsync', () => {
    it('supports FormData', async () => {
      const example = new FormData();
      example.set('hello-world', new File(['Hello World'], 'hello.txt', {
        type: 'text/plain',
        lastModified: 1681027542680,
      }));
      example.set('foo-bar', new File(['Foo Bar'], 'foo-bar.txt', {
        type: 'text/plain',
        lastModified: 1681027542680,
      }));
      const result = await serializeAsync(Promise.resolve(example));
      expect(result).toMatchSnapshot();
      const back = await deserialize<Promise<FormData>>(result);
      expect(back).toBeInstanceOf(FormData);
    });
  });
  describe('toJSON', () => {
    it('supports FormData', () => {
      const example = new FormData();
      example.set('hello', 'world');
      example.set('foo', 'bar');
      const result = toJSON(example);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<FormData>(result);
      expect(back).toBeInstanceOf(FormData);
    });
  });
  describe('toJSONAsync', () => {
    it('supports FormData', async () => {
      const example = new FormData();
      example.set('example', new File(['Hello World'], 'hello.txt', {
        type: 'text/plain',
        lastModified: 1681027542680,
      }));
      example.set('foo-bar', new File(['Foo Bar'], 'foo-bar.txt', {
        type: 'text/plain',
        lastModified: 1681027542680,
      }));
      const result = await toJSONAsync(Promise.resolve(example));
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = await fromJSON<Promise<FormData>>(result);
      expect(back).toBeInstanceOf(FormData);
      expect(String(back)).toBe(String(example));
    });
  });
  describe('crossSerialize', () => {
    it('supports FormData', () => {
      const example = new FormData();
      example.set('hello', 'world');
      example.set('foo', 'bar');
      const result = crossSerialize(example);
      expect(result).toMatchSnapshot();
    });
    describe('scoped', () => {
      it('supports FormData', () => {
        const example = new FormData();
        example.set('hello', 'world');
        example.set('foo', 'bar');
        const result = crossSerialize(example, { scopeId: 'example' });
        expect(result).toMatchSnapshot();
      });
    });
  });
  describe('crossSerializeAsync', () => {
    it('supports FormData', async () => {
      const example = new FormData();
      example.set('hello-world', new File(['Hello World'], 'hello.txt', {
        type: 'text/plain',
        lastModified: 1681027542680,
      }));
      example.set('foo-bar', new File(['Foo Bar'], 'foo-bar.txt', {
        type: 'text/plain',
        lastModified: 1681027542680,
      }));
      const result = await crossSerializeAsync(Promise.resolve(example));
      expect(result).toMatchSnapshot();
    });
    describe('scoped', () => {
      it('supports FormData', async () => {
        const example = new FormData();
        example.set('hello-world', new File(['Hello World'], 'hello.txt', {
          type: 'text/plain',
          lastModified: 1681027542680,
        }));
        example.set('foo-bar', new File(['Foo Bar'], 'foo-bar.txt', {
          type: 'text/plain',
          lastModified: 1681027542680,
        }));
        const result = await crossSerializeAsync(Promise.resolve(example), { scopeId: 'example' });
        expect(result).toMatchSnapshot();
      });
    });
  });
  describe('crossSerializeStream', () => {
    it('supports FormData', async () => new Promise<void>((done) => {
      const example = new FormData();
      example.set('hello', 'world');
      example.set('foo', 'bar');
      crossSerializeStream(Promise.resolve(example), {
        onSerialize(data) {
          expect(data).toMatchSnapshot();
        },
        onDone() {
          done();
        },
      });
    }));
    describe('scoped', () => {
      it('supports FormData', async () => new Promise<void>((done) => {
        const example = new FormData();
        example.set('hello', 'world');
        example.set('foo', 'bar');
        crossSerializeStream(Promise.resolve(example), {
          scopeId: 'example',
          onSerialize(data) {
            expect(data).toMatchSnapshot();
          },
          onDone() {
            done();
          },
        });
      }));
    });
  });
});
