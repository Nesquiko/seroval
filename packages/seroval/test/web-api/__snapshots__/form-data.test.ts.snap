// Vitest Snapshot v1

exports[`FormData > serialize > supports FormData 1`] = `"(h=>((h=new FormData(),h.append(\\"hello\\",\\"world\\"),h.append(\\"foo\\",\\"bar\\"),h)))()"`;

exports[`FormData > serializeAsync > supports FormData 1`] = `"(h=>(Promise.resolve((h=new FormData(),h.append(\\"hello-world\\",new File([new Uint8Array([72,101,108,108,111,32,87,111,114,108,100]).buffer],\\"hello.txt\\",{type:\\"text/plain\\",lastModified:1681027542680})),h.append(\\"foo-bar\\",new File([new Uint8Array([70,111,111,32,66,97,114]).buffer],\\"foo-bar.txt\\",{type:\\"text/plain\\",lastModified:1681027542680})),h))))()"`;

exports[`FormData > toJSON > supports FormData 1`] = `"{\\"t\\":{\\"t\\":26,\\"i\\":0,\\"d\\":{\\"k\\":[\\"hello\\",\\"foo\\"],\\"v\\":[{\\"t\\":1,\\"s\\":\\"world\\"},{\\"t\\":1,\\"s\\":\\"bar\\"}],\\"s\\":2}},\\"r\\":0,\\"i\\":false,\\"f\\":16383,\\"m\\":[]}"`;

exports[`FormData > toJSONAsync > supports FormData 1`] = `"{\\"t\\":{\\"t\\":12,\\"i\\":0,\\"f\\":{\\"t\\":26,\\"i\\":1,\\"d\\":{\\"k\\":[\\"example\\",\\"foo-bar\\"],\\"v\\":[{\\"t\\":24,\\"i\\":2,\\"c\\":\\"text/plain\\",\\"m\\":\\"hello.txt\\",\\"f\\":{\\"t\\":21,\\"i\\":3,\\"s\\":[72,101,108,108,111,32,87,111,114,108,100]},\\"b\\":1681027542680},{\\"t\\":24,\\"i\\":4,\\"c\\":\\"text/plain\\",\\"m\\":\\"foo-bar.txt\\",\\"f\\":{\\"t\\":21,\\"i\\":5,\\"s\\":[70,111,111,32,66,97,114]},\\"b\\":1681027542680}],\\"s\\":2}}},\\"r\\":0,\\"i\\":false,\\"f\\":16383,\\"m\\":[]}"`;
